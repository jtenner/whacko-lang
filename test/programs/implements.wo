export interface MyInterface {
  field: i32;
  method(a: i64): i32;
}

export class MyClass implements MyInterface {
  field: i32;
  constructor() {
    this.field = i32(42);
  }
  method(a: i64): i32 {
    this.field = i32(a);
    return i32(2);
  }
}

export class MyOtherClass implements MyInterface {
  field: i32;
  constructor() {
    this.field = i32(42);
  }
  method(a: i64): i32 {
    this.field = i32(a);
    return i32(1);
  }
}

export fn useMyInterface(a: MyInterface): i32 {
  a.field = i32(0);
  return a.method(69);
}

export fn _start(): void {
  let a = new MyClass();
  let b = new MyOtherClass();
  useMyInterface(a);
}
