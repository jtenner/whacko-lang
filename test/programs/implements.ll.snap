; ModuleID = 'whacko'
source_filename = "whacko"
target triple = "wasm32-wasi"

declare void @__whacko_gc_initialize()

declare ptr @__whacko_gc_alloc(i32, i32)

declare void @__whacko_gc_store_field_ptr(i32, ptr, ptr, ptr)

define void @_start() {
entry:
  call void @__whacko_gc_initialize()
  %"alloca~0" = alloca ptr, align 8
  %"alloca~1" = alloca ptr, align 8
  %"result~423" = call ptr @__whacko_gc_alloc(i32 1, i32 ptrtoint (ptr getelementptr ({ ptr, ptr, i32, i32, i32 }, ptr null, i32 1) to i32))
  call void @"callable:[test/programs/implements.wo]MyClass.constructor;methodType:classType:[test/programs/implements.wo]MyClass<>#()=>voidType"(ptr %"result~423")
  store ptr %"result~423", ptr %"alloca~0", align 8
  %"result~425" = call ptr @__whacko_gc_alloc(i32 2, i32 ptrtoint (ptr getelementptr ({ ptr, ptr, i32, i32, i32 }, ptr null, i32 1) to i32))
  call void @"callable:[test/programs/implements.wo]MyOtherClass.constructor;methodType:classType:[test/programs/implements.wo]MyOtherClass<>#()=>voidType"(ptr %"result~425")
  store ptr %"result~425", ptr %"alloca~1", align 8
  %"339~inst" = load ptr, ptr %"alloca~0", align 8
  %"343~inst" = call i32 @useMyInterface(ptr %"339~inst")
  ret void
}

define void @"callable:[test/programs/implements.wo]MyClass.constructor;methodType:classType:[test/programs/implements.wo]MyClass<>#()=>voidType"(ptr %0) {
entry:
  %"alloca~0" = alloca ptr, align 8
  store ptr %0, ptr %"alloca~0", align 8
  %"gep~426" = getelementptr { ptr, ptr, i32, i32, i32 }, ptr %"alloca~0", i32 4
  %"372~inst" = load i32, ptr %"gep~426", align 4
  %"gep~427" = getelementptr { ptr, ptr, i32, i32, i32 }, ptr %"alloca~0", i32 4
  store i32 42, ptr %"gep~427", align 4
  ret void
}

define void @"callable:[test/programs/implements.wo]MyOtherClass.constructor;methodType:classType:[test/programs/implements.wo]MyOtherClass<>#()=>voidType"(ptr %0) {
entry:
  %"alloca~0" = alloca ptr, align 8
  store ptr %0, ptr %"alloca~0", align 8
  %"gep~428" = getelementptr { ptr, ptr, i32, i32, i32 }, ptr %"alloca~0", i32 4
  %"363~inst" = load i32, ptr %"gep~428", align 4
  %"gep~429" = getelementptr { ptr, ptr, i32, i32, i32 }, ptr %"alloca~0", i32 4
  store i32 42, ptr %"gep~429", align 4
  ret void
}

define i32 @useMyInterface(ptr %0) {
entry:
  %"alloca~0" = alloca ptr, align 8
  store ptr %0, ptr %"alloca~0", align 8
  %"trampolineCall~430" = call ptr @"interfaceType:[test/programs/implements.wo]MyInterface<>#field"(ptr %"alloca~0")
  %"346~inst" = load i32, ptr %"trampolineCall~430", align 4
  %"trampolineCall~431" = call ptr @"interfaceType:[test/programs/implements.wo]MyInterface<>#field"(ptr %"alloca~0")
  store i32 0, ptr %"trampolineCall~431", align 4
  %"353~inst" = load ptr, ptr %"alloca~0", align 8
  %"360~inst" = call i32 @"callable:[test/programs/implements.wo]MyInterface.method;methodType:interfaceType:[test/programs/implements.wo]MyInterface<>#(i64Type)=>i32Type"(ptr %"353~inst", i64 69)
  ret i32 %"360~inst"
}

define i32 @"callable:[test/programs/implements.wo]MyInterface.method;methodType:interfaceType:[test/programs/implements.wo]MyInterface<>#(i64Type)=>i32Type"(ptr %0, i64 %1) {
"entry~441":
  %"typeGEP~442" = getelementptr { ptr, ptr, i32, i32 }, ptr %0, i32 2
  %2 = load i32, ptr %"typeGEP~442", align 4
  switch i32 %2, label %"default~443" [
    i32 1, label %"callable:[test/programs/implements.wo]MyClass.method;methodType:classType:[test/programs/implements.wo]MyClass<>#(i64Type)=>i32Type"
    i32 2, label %"callable:[test/programs/implements.wo]MyOtherClass.method;methodType:classType:[test/programs/implements.wo]MyOtherClass<>#(i64Type)=>i32Type"
  ]

"default~443":                                    ; preds = %"entry~441"
  unreachable

"callable:[test/programs/implements.wo]MyClass.method;methodType:classType:[test/programs/implements.wo]MyClass<>#(i64Type)=>i32Type": ; preds = %"entry~441"
  %3 = call i32 @"callable:[test/programs/implements.wo]MyClass.method;methodType:classType:[test/programs/implements.wo]MyClass<>#(i64Type)=>i32Type"(ptr %0, i64 %1)
  ret i32 %3

"callable:[test/programs/implements.wo]MyOtherClass.method;methodType:classType:[test/programs/implements.wo]MyOtherClass<>#(i64Type)=>i32Type": ; preds = %"entry~441"
  %4 = call i32 @"callable:[test/programs/implements.wo]MyOtherClass.method;methodType:classType:[test/programs/implements.wo]MyOtherClass<>#(i64Type)=>i32Type"(ptr %0, i64 %1)
  ret i32 %4
}

define i32 @"callable:[test/programs/implements.wo]MyClass.method;methodType:classType:[test/programs/implements.wo]MyClass<>#(i64Type)=>i32Type"(ptr %0, i64 %1) {
entry:
  %"alloca~0" = alloca ptr, align 8
  store ptr %0, ptr %"alloca~0", align 8
  %"alloca~1" = alloca i64, align 8
  store i64 %1, ptr %"alloca~1", align 4
  %"gep~432" = getelementptr { ptr, ptr, i32, i32, i32 }, ptr %"alloca~0", i32 4
  %"404~inst" = load i32, ptr %"gep~432", align 4
  %"405~inst" = load i64, ptr %"alloca~1", align 4
  %"409~inst" = trunc i64 %"405~inst" to i32
  %"gep~433" = getelementptr { ptr, ptr, i32, i32, i32 }, ptr %"alloca~0", i32 4
  store i32 %"409~inst", ptr %"gep~433", align 4
  ret i32 2
}

define i32 @"callable:[test/programs/implements.wo]MyOtherClass.method;methodType:classType:[test/programs/implements.wo]MyOtherClass<>#(i64Type)=>i32Type"(ptr %0, i64 %1) {
entry:
  %"alloca~0" = alloca ptr, align 8
  store ptr %0, ptr %"alloca~0", align 8
  %"alloca~1" = alloca i64, align 8
  store i64 %1, ptr %"alloca~1", align 4
  %"gep~434" = getelementptr { ptr, ptr, i32, i32, i32 }, ptr %"alloca~0", i32 4
  %"390~inst" = load i32, ptr %"gep~434", align 4
  %"391~inst" = load i64, ptr %"alloca~1", align 4
  %"395~inst" = trunc i64 %"391~inst" to i32
  %"gep~435" = getelementptr { ptr, ptr, i32, i32, i32 }, ptr %"alloca~0", i32 4
  store i32 %"395~inst", ptr %"gep~435", align 4
  ret i32 1
}

define ptr @"interfaceType:[test/programs/implements.wo]MyInterface<>#field"(ptr %0) {
"entry~436":
  %"typeGEP~437" = getelementptr { ptr, ptr, i32, i32 }, ptr %0, i32 2
  %1 = load i32, ptr %"typeGEP~437", align 4
  switch i32 %1, label %"default~438" [
    i32 1, label %"classType:[test/programs/implements.wo]MyClass<>"
    i32 2, label %"classType:[test/programs/implements.wo]MyOtherClass<>"
  ]

"default~438":                                    ; preds = %"entry~436"
  unreachable

"classType:[test/programs/implements.wo]MyClass<>": ; preds = %"entry~436"
  %"resultGEP~439" = getelementptr { ptr, ptr, i32, i32, i32 }, ptr %0, i32 2
  ret ptr %"resultGEP~439"

"classType:[test/programs/implements.wo]MyOtherClass<>": ; preds = %"entry~436"
  %"resultGEP~440" = getelementptr { ptr, ptr, i32, i32, i32 }, ptr %0, i32 2
  ret ptr %"resultGEP~440"
}

declare void @__whacko_gc_visit(ptr)

define void @__visit_whacko_object(ptr %0) {
"entry~444":
  %"typeGEP~446" = getelementptr { ptr, ptr, i32, i32 }, ptr %0, i32 2
  %1 = load i32, ptr %"typeGEP~446", align 4
  switch i32 %1, label %"unreachable~445" [
    i32 1, label %"classType:[test/programs/implements.wo]MyClass<>"
    i32 2, label %"classType:[test/programs/implements.wo]MyOtherClass<>"
    i32 3, label %"classType:[std/String.wo]String<>"
  ]

"unreachable~445":                                ; preds = %"entry~444"
  unreachable

"classType:[test/programs/implements.wo]MyClass<>": ; preds = %"entry~444"
  ret void

"classType:[test/programs/implements.wo]MyOtherClass<>": ; preds = %"entry~444"
  ret void

"classType:[std/String.wo]String<>":              ; preds = %"entry~444"
  ret void
}
